### ISSUE TITLE
Add support for additional configuration of consul/vault/nomad before startup

### ISSUE THREAD

**Issue by fredrikhgrelland** [fredrikhgrelland/vagrant-hashistack#81]
While using the vagrant box you might want or need to change or add some configuration of the underlying hashistack. We should allow for putting configuration files in a directory and automatically add them to the appropriate directories on the host before starting the hashistack.

**Comment by m88614** [comment]
I strongly support this. I started symlinking stuff literally half an hour ago.

**Comment by fredrikhgrelland** [comment]
> I strongly support this. I started symlinking stuff literally half an hour ago.

Hehe, I am trying to change nomad config to allow for plugin in template and I realised we had no support for that 🍭

**Comment by m88614** [comment]
> Hehe, I am trying to change nomad config to allow for plugin in template and I realised we had no support for that

And I'm fiddling with the startup-playbooks. 

Atm I see many benefits of having everything that needs to be copied onto specific places on the machine being symlinks instead. 
IE this: https://github.com/fredrikhgrelland/vagrant-hashistack/blob/aacb7acbd3c7947c35f6cee7ed34eccd7c08e1bb/ansible/playbook.yml#L16-L19 would just make a symlink instead.

I've manually symlinked it at this point, which gives me the full range of my actual development environment, fused together with the environment that the hashistack box provides. It's actually pretty damn cool

**Comment by zhenik** [comment]
Does the action of copying file not enough to resolve this issue?

## Example
`config-override.hcl`
```hcl
acl {
  default_policy = "deny"
  down_policy = "extend-cache"
  enabled = true
  tokens {
    master = "master"
  }
}
```

`Vagrantfile`
```
Vagrant.configure("2") do |config|
    # some config
   
    # copy own config ~> config merge feature
    config.vm.provision "file", source: "config-override.hcl", destination: "/etc/consul.d/config-override.hcl"

    config.vm.provision "ansible_local" do |startup|
        run = "always"
        startup.playbook = "/etc/ansible/startup.yml"
    end
end
```

### Feature: config merge
At consul start, directory `/etc/consul.d/config-override.hcl` will contain two config files `config.hcl` and `config-override.hcl`,
configuration files will be merged in [lexical order](https://www.consul.io/docs/agent/options.html): `config.hcl` and then`config-override.hcl`
### Feature: new config
The user could provide new config, with destination: "/etc/consul.d/config.hcl"
```
  # copy own config ~> config merge feature
    config.vm.provision "file", source: "config-new.hcl", destination: "/etc/consul.d/config.hcl"
```

**Comment by m88614** [comment]
My use case atm: a yml file that I'd like to test in the vagrant box's environment (with vault, TF, etc.). I go into the box and run it with `ansible-playbook playbook.yml` to get immediate feedback. With a copy I would have to run the copy command to replace the old one with the new changes every time I make a minute change. With a symlink I can have the file open in my VDI's editor, make changes, then immediately go to the box and run it

**Comment by fredrikhgrelland** [comment]
> Does the action of copying file not enough to resolve this issue?
> 
> ## Example
> `config-override.hcl`
> 
> ```terraform
> acl {
>   default_policy = "deny"
>   down_policy = "extend-cache"
>   enabled = true
>   tokens {
>     master = "master"
>   }
> }
> ```
> 
> `Vagrantfile`
> 
> ```
> Vagrant.configure("2") do |config|
>     # some config
>    
>     # copy own config ~> config merge feature
>     config.vm.provision "file", source: "config-override.hcl", destination: "/etc/consul.d/config-override.hcl"
> 
>     config.vm.provision "ansible_local" do |startup|
>         run = "always"
>         startup.playbook = "/etc/ansible/startup.yml"
>     end
> end
> ```
> 
> ### Feature: config merge
> At consul start, directory `/etc/consul.d/config-override.hcl` will contain two config files `config.hcl` and `config-override.hcl`,
> configuration files will be merged in [lexical order](https://www.consul.io/docs/agent/options.html): `config.hcl` and then`config-override.hcl`
> 
> ### Feature: new config
> The user could provide new config, with destination: "/etc/consul.d/config.hcl"
> 
> ```
>   # copy own config ~> config merge feature
>     config.vm.provision "file", source: "config-new.hcl", destination: "/etc/consul.d/config.hcl"
> ```

Yes! This exactly does it for me. This makes it a simple documentation addition. Brilliant.

**Comment by fredrikhgrelland** [comment]
> My use case atm: a yml file that I'd like to test in the vagrant box's environment (with vault, TF, etc.). I ssh into the box and run it with `ansible-playbook playbook.yml` to get immediate feedback. With a copy I would have to run the copy command to replace the old one with the new changes every time I make a minute change. With a symlink I can have the file open in my VDI's editor, make changes, then immediately go to the box and run it
> 
> edit: I'm open to being way off, your post was a proper TL:DR for me atm, I'm taking the rest of the evening off :D

I am hearing you. We should have a default directory where ansible imports a playbook. pre or post startup. That is a different Issue. So let us make it as such.

**Comment by zhenik** [comment]
But why we need to provide pre, post startups **directory**? 

I think users can expand `Vagrantfile` as  they wish, with additional changes (provisioners, providers, etc...)
I would suggest the approach is to make documentation clear about it and provide `Vagrantfile` examples (with proper testing).

## Example
```
Vagrant.configure("2") do |config|
    # some config
   
    # PRESTARTUP  (copy some files, override config, additional provisioning, etc...)

    # STARTUP
    config.vm.provision "ansible_local" do |startup|
        run = "always"
        startup.playbook = "/etc/ansible/startup.yml"
    end

    # POSTSTARTUP (copy some files, additional provisioning, etc...)

end
```

**Comment by fredrikhgrelland** [comment]
> But why we need to provide pre, post startups **directory**?

I think you are right. **period**

Let us package standard configuration flavors and document the properly.

**Comment by fredrikhgrelland** [comment]
closes #82

**Comment by fredrikhgrelland** [comment]
Note:
It will not allow using the file provisioner as it can not add that to a dir owned by root see: https://stackoverflow.com/questions/54067192/vagrant-config-vm-provision-does-not-allow-me-to-copy-a-file-to-etc-nginx-conf

solution is to reference something in the share directory `/vagrant` using the shell provider:
```
config.vm.provision "shell", inline: "cp /vagrant/conf/nomad/config-override.hcl /etc/nomad.d/config-override.hcl"
```
